docs/source/tutorials/chat.rst

```diff
@@ -0,0 +1,164 @@
+      * How to use your own chat dataset to fine-tune LLaMA3
+
+    .. grid-item-card:: :octicon:`list-unordered;1em;` Prerequisites
+
+      * Be familiar with :ref:`configuring datasets<dataset_tutorial_label>`
+      * Know how to :ref:`download LLaMA3 weights <llama3_label>`
+
+
+Prompt templates & special tokens
+---------------------------------
+
+Prompt templates and tokenization schemes are often conflated, and it's sometimes
+unclear what you need to do to format your data to optimize the training performance
+of your model. Let's walk through the LLaMA2/LLaMA3 templates to better understand
+the distinction.
+
+In torchtune, a prompt template is simply a structured way to format your prompt
+into the model. The prompt template adds flavor text to prime the model to expect
+a certain structure when it is used for inference. From the `official LLaMA2 prompt
+template guide <https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-2>`_
+for the LLaMA2 chat model, we can see that special tags are added:
+
+.. code-block:: text
+
+    <s>[INST] <<SYS>>
+    {{ system_prompt }}
+    <</SYS>>
+
+    {{ user_message_1 }} [/INST] {{ model_answer_1 }} </s>
+    <s>[INST] {{ user_message_2 }} [/INST]
+
+Since the chat model was pretrained with this prompt template, if you want to run
+inference on the model, you'll need to use the same template for optimal performance
+on chat data. Otherwise, the model will just perform standard text completion, which
+may or may not align with your intended use case.
+
+If you look at our :class:`~torchtune.data.Llama2ChatFormat` class, you'll notice that
+we don't include the :code:`<s>` and :code:`</s>` tokens. These are the beginning-of-sequence
+(BOS) and end-of-sequence (EOS) tokens that are represented differently in the tokenizer
+than the rest of the prompt template. Let's tokenize some examples with the
+:class:`~torchtune.modules.tokenizers.SentencePieceTokenizer` used by LLaMA2 to see
+why.
+
+.. code-block:: python
+
+    from torchtune.modules.tokenizers import SentencePieceTokenizer
+
+    tokenizer = SentencePieceTokenizer("/tmp/Llama-2-7b-hf/tokenizer.model")
+    text = "[INST] <<SYS>> Hello"
+    tokens = tokenizer.encode(text, add_bos=True, add_eos=True)
+    print(tokens)
+    # [1, 518, 25580, 29962, 3532, 14816, 29903, 6778, 15043, 2]
+
+We've added the BOS and EOS tokens when encoding our example text. This shows up
+as IDs 1 and 2. We can verify that these are our BOS and EOS tokens.
+
+.. code-block:: python
+
+    print(tokenizer.spm_model.piece_to_id("<s>"))
+    # 1
+    print(tokenizer.spm_model.piece_to_id("</s>"))
+    # 2
+
+The BOS and EOS tokens are what we call special tokens, because they have their own
+reserved token IDs. This means that they will index to their own individual vectors in
+the model's learnt embedding table. The rest of the prompt template tags, :code:`[INST]`
+and :code:`<<SYS>>` are tokenized as normal text and not their own IDs.
```

Comment:
Is the point here that `[INST]` and `<<SYS>>` are not special tokens? If so, just say that (though tbh I wouldn't even go into it, and would just make this section about tokenization and special tokens)

Reply:
Yes, but I would like to show that they are tokenized differently to illustrate my point, so I need to talk about it in the context of tolenization